### PR TITLE
BUG: Fix dtypes order when ordering is different from original file in pandas.io.stata.read_stata

### PR DESCRIPTION
**Pull request by jbuyl**
This fixes http://stackoverflow.com/questions/31783392/pandas-adding-columns-to-filter-will-mess-with-data-structure
The dtypes of the DataFrame returned by `read_stata` have the same order as the original stata file, even if `columns` are specified with a different order.

### CODE REVIEW

File: pandas/io/tests/test_stata.py [hunk 1]
```diff
@@ -780,8 +780,14 @@ def test_drop_column(self):
         expected = expected[columns]
         dropped = read_stata(self.dta15_117, convert_dates=True,
                              columns=columns)
-
         tm.assert_frame_equal(expected, dropped)
+
+        columns = ['int_', 'long_', 'byte_']
+        expected = expected[columns]
+        reordered = read_stata(self.dta15_117, convert_dates=True,
+                               columns=columns)
```
Comment (jreback) [inline pandas/io/tests/test_stata.py]: can you tests with an invalid column? (not exactly sure what we should do though), could reindex, ``KeyError`` or ignore.

Comment (jbuyl) [inline pandas/io/tests/test_stata.py]: There are tests on invalid columns in this test function already just a bit
below this line. I just added the case of columns with different order.

File: pandas/io/stata.py [hunk 2]
```diff
@@ -1615,13 +1615,14 @@ def _do_select_columns(self, data, columns):
...
-                if col in column_set:
-                    matched.update([col])
-                    dtyplist.append(self.dtyplist[i])
-                    typlist.append(self.typlist[i])
-                    fmtlist.append(self.fmtlist[i])
-                    lbllist.append(self.lbllist[i])
+            data_column_list = data.columns.tolist()
+            for col in columns:
+                matched.update([col])
+                i = data_column_list.index(col)
```
Comment (jreback) [inline pandas/io/stata.py]: use ``data.columns.get_loc(col)``

**Comment by jreback** [comment]
looks good.

- minor code change
- pls add a note in whatsnew
- squash

**Comment by jbuyl** [comment]
Done, I got rid of the `matched` dict() as well because it doesn't do anything.

**Comment by jreback** [comment]
lgtm

pls rebase and squash (FYI move the release note higher up in the bug fix list and you wont get conflicts)

**Comment by jbuyl** [comment]
Is this right? Looks like I screwed up something

**Comment by jreback** [comment]
git help here: http://pandas.pydata.org/pandas-docs/stable/contributing.html

basically do

got rebase -i origin/master

**Comment by jbuyl** [comment]
Right, well I pretty much did that. How does the PR look?

**Comment by jreback** [comment]
did you ``git push <username> <branchname> -f``. You have 16 commits here. You should have 1.

**Comment by jbuyl** [comment]
This is a mess, I'm just going to open a new PR for this

**Pull request closed without merging**